internal/controllers/httpsedge_controller.go

```diff
@@ -209,51 +207,74 @@ func (r *HTTPSEdgeReconciler) reconcileRoutes(ctx context.Context, edge *ingress
 
 		match := r.getMatchingRouteFromEdgeStatus(edge, routeSpec)
 		var route *ngrok.HTTPSEdgeRoute
+		// Now we go ahead and create the route if it doesn't exist, or find the existing route.
+		// It's important to note here that we are intentionally ommiting the `route.Backend` for new routes, and
+		//  removing it from existing routes. The success or failure of applying a route's modules is then strongly 
+		//  linked the state of its backend. Thus, any route with a backend is considered properly configured.
+		//  See https://github.com/ngrok/kubernetes-ingress-controller/issues/208 for additional context.
 		if match == nil {
-			r.Log.Info("Creating new route", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType, "backendID", backend.ID)
-			// This is a new route, so we need to create it
+			r.Log.Info("Creating new route", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType)
 			req := &ngrok.HTTPSEdgeRouteCreate{
 				EdgeID:    edge.Status.ID,
 				Match:     routeSpec.Match,
 				MatchType: routeSpec.MatchType,
-				Backend: &ngrok.EndpointBackendMutate{
-					BackendID: backend.ID,
-				},
 			}
-			route, err = r.NgrokClientset.HTTPSEdgeRoutes().Create(ctx, req)
+			route, err = edgeRoutes.Create(ctx, req)
 		} else {
-			r.Log.Info("Updating route", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType, "backendID", backend.ID)
-			// This is an existing route, so we need to update it
+			r.Log.Info("Route found, disabling backend to prepare for module updates", "edgeID", edge.Status.ID, "match", routeSpec.Match, "matchType", routeSpec.MatchType)
 			req := &ngrok.HTTPSEdgeRouteUpdate{
 				EdgeID:    edge.Status.ID,
 				ID:        match.ID,
 				Match:     routeSpec.Match,
 				MatchType: routeSpec.MatchType,
-				Backend: &ngrok.EndpointBackendMutate{
-					BackendID: backend.ID,
-				},
```

Comment:
For my clarification, so if you have an Edge with 4 routes and you make any update to the edge, its going to go through and remove the backend for each route temporarily, even if there was no need to update it correct? It would be nice to be able to skip any routes that are already matching what is in the backend API to avoid blipping each endpoint on/off on any change.

I think the data is available via the edge api, but it might be difficult to do all the diffing and would be better done with an atomic API server side.

Reply:
> if you have an Edge with 4 routes and you make any update to the edge, its going to go through and remove the backend for each route temporarily, even if there was no need to update it correct?

That is correct in the current set of changes.

> It would be nice to be able to skip any routes that are already matching what is in the backend API to avoid blipping each endpoint on/off on any change.

Absolutely agree, I'd love to have something that determines if there is a diff to actually apply. I tried to encapsulate that notion in my notes at the bottom of the description (something like `planModuleUpdates`).

> I think the data is available via the edge api, but it might be difficult to do all the diffing and would be better done with an atomic API server side.

Hmm, okay. Where would you think to put it on the server side then?

Reply:
I'm wondering if we should only do this in this PR for creates based on this statement as the change for updates was an additional portion of my original comment and was an attempt to try and handle the lack of an atomic api.

This means that:

- Creates: No Change
- Update: The way it worked previously (edge is alive even on error)

But, ultimately, Alex is correct that uptime and user-experience where the end-user only experiences an outage from their own mistakes is seemingly better than an enforced one during update.

---

For the future, there could be a few other approaches we could take that don't risk the user-experience without the lack of an atomic api today:

- Create staging-edge with new configuration for validation that things work as intended
  - On success
    - Delete staging-edge
    - Run workflow to create / update existing edges
  - On failure
    - Delete staging-edge
    - Error out

Open questions for this:
- Any limitations around number of objects / billing on additional objects
- If so how to omit from usage

But this would be able to solve not having validation on the client and the create/update issue without atomic api

Will open a new issue for the above, there were alternatives but the downsides were significant:

- Blue-green swapping from staging-edge to production-edge
  - Hostname swapping with downtime
  - Labels swapping
  - Sync issues? Never tried it so not sure how the system would handle live agents when this happens
- Route swapping but routes are limited per account (3-5?) so this would cause issues and is harder to remove that limit

Reply:
> we are trading user experience for security which doesn't feel great when thinking about usage

The UX feels quite bad, but I think the security issues are worse. If someone's using a security sensitive module, like oauth, ip policies, etc, and we start routing through unauthorized requests, that feels worse than the ingress controller not existing at all.

> forced downtimes can cause interruptions and the shuttering of the service could have greater impacts than a partial one and a partial one would be expected with invalid configuration values.

Forced downtime is still better than a partial outage where the outage is "oauth is no longer being applied, all requests to /admin go through" for example.

It's important to note that completely valid configuration changes will fail to update with how it is written currently, so this doesn't only apply to "invalid configuration values". See https://github.com/ngrok/kubernetes-ingress-controller/issues/221 for a couple examples of these.

------

> Create staging-edge with new configuration for validation that things work as intended
> On success
> * Delete staging-edge
> * Run workflow to create / update existing edges

This doesn't ensure that the transition from the previous edge to the new edge is valid, i.e. for the items mentioned in #221, the new edge would create, but the update still may not be valid.

If we follow this path, I think a simpler pattern is the one you mention below of doing blue/green swaps:

1. Create a new edge with the new configuration
2. Delete the hostport from the old edge
3. Add the hostport to the new edge
4. Delete the old edge

If we're okay having two edges (an "active" and "staging" one), we might as well just shuffle the host port over rather than delete the new one and try to transition.

This solution also fixes the update transition issues since it lets us treat edges as entirely immutable once they're "active". This one seems pretty compelling.